### DISCUSSION TITLE
Save to Folder not including definition.pbism and .platform (for TMDL)

### DISCUSSION THREAD

**Discussion by martinmolm-elkjop** [TabularEditor/TabularEditor3#1418] [🙋♀️TE3 Q&A] [answered]
### Description

When Saving a (Published) Model to Folder (in TMDL-format), it is not saving the definition.pbism or the .platform files.
I assume this is because you've based it on the C# Library and not the REST API. 
I've noticed that the Fabric Rest API call to get the Semantic Model Definition is able to return these files. So, could you please use your influence at Microsoft to have them complete their C# Library/API, and/or modify your code to utilize the REST API? The .platform file is rather critical to get when you're saving a model I think.

### Tabular Editor 3 Version

3.18.2

### Steps to Reproduce

Connect to any model in your service.
Make sure to "Save to Folder" in TMDL format.
Notice the missing .platform file, definition.pbism and .definition-folder.

### Expected behavior

It should save in a .SemanticModel-folder including the  .platform file, definition.pbism and .definition-folder.

### Windows Version

11

**Comment by otykier** [comment]
Thanks for reaching out. 

It's a bit unclear to me what you're trying to achieve. Under normal circumstances, it shouldn't be necessary to modify the .pbism or .platform files (or any of the other non-TMDL metadata files) when making changes to a model.

Based on your repro steps, it sounds like you're trying to perform a full "export" of your Power BI / Fabric workspace (or at least of a particular semantic model within the workspace). That's not what Tabular Editor is designed for. There are better ways to do that, such as by cloning the git repo locally, or by using the REST API as you suggest.

Using the REST API is not an option for us, as that requires separate authentication, and users who have access to the model metadata through the XMLA endpoint, might not have REST API access (or REST API access may be completely disabled at an organization level).

**Comment by martinmolm-elkjop** [comment]
Thanks for your quick response Daniel.

I'll try to elaborate a bit on what I'm trying to achieve.

In my scenario I want to be able to create a Model using Tabular Editor, and then publish it using Git, rather than publishing it directly to a workspace. That would entail saving the model to folder (my local branch repository) and then commit to branch (and alternatively sync branch to connected workspace, or merge to main).
The benefits of this would be the ability to specify GitHub Actions on commit (run BPA for instance) or other steps to do QA. It would also remove the need for specific GitHub Integrated Branch-Workspaces (which I find quite cumbersome to set up automatically).

The point I guess I'm trying to make is that if you start to create a Model in Tabular Editor it would save a suboptimal representation of the Model according to the TMDL specifications 
...if your goal is to publish the model to the service in any other way than through the XMLA endpoint that is... 

I really appreciate your thoughts on this, I might be way off in my thinking here...

**Comment by otykier** [comment, marked as answer]
I see, thanks for clarifying.

Our official recommendation, if you want to deploy the model to the Fabric workspace initially through Git, rather than through the XMLA endpoint, is to create a dummy model in PBI Desktop, and save that locally as a PBIP. This will then create the model metadata files including all the supporting files (see [step 3, option 3 in this blog post](https://blog.tabulareditor.com/2024/04/11/tabular-editor-and-fabric-git-integration/)), which you can then commit/push. Upon updating the Fabric workspace based on the git content, the new model will be picked up and created in the workspace.

That being said, you can actually create the supporting files yourself. It's not documented anywhere, though, so it's probably not a technique supported by Microsoft. For a model that uses the TMDL format, you only need the following:

- **definition/** - a folder containing all the .tmdl files created by Tabular Editor when saving the model
- **definition.pbism** - a JSON file containing the following code:
  - `{ "version": "4.0", "settings": {} }`
- **.platform** - a JSON file containing the below code:

```json
{
  "$schema": "https://developer.microsoft.com/json-schemas/fabric/gitIntegration/platformProperties/2.0.0/schema.json",
  "metadata": {
    "type": "SemanticModel",
    "displayName": "<ModelName>"
  },
  "config": {
    "version": "2.0",
    "logicalId": "<Guid>"
  }
}
```

1. Replace \<ModelName\> with the name you want the model to have in the Fabric workspace UI.
2. Replace \<Guid\> with any random GUID.

**Comment by mlonsk** [comment]
I will move this thread to our discussion section, as it is not an issue with Tabular Editor, but contains information that could be relevant to others.
